lib/gruf/hooks/registry.rb

```diff
@@ -0,0 +1,159 @@
+
+      ##
+      # Remove an hook from the registry
+      #
+      # @param [Class] hook_class The hook class to remove
+      # @raise [HookNotFoundError] if the hook is not found
+      #
+      def remove(hook_class)
+        pos = @registry.find_index { |opts| opts.fetch(:klass, '') == hook_class }
+        raise HookNotFoundError if pos.nil?
+
+        @registry.delete_at(pos)
+      end
+
+      ##
+      # Insert an hook before another specified hook
+      #
+      # @param [Class] before_class The hook to insert before
+      # @param [Class] hook_class The class of the hook to add
+      # @param [Hash] options A hash of options to pass into the hook during initialization
+      # @raise [HookNotFoundError] if the before hook is not found
+      #
+      def insert_before(before_class, hook_class, options = {})
+        hooks_mutex do
+          pos = @registry.find_index { |opts| opts.fetch(:klass, '') == before_class }
+          raise HookNotFoundError if pos.nil?
+
+          @registry.insert(
+            pos,
+            klass: hook_class,
+            options: options
+          )
+        end
+      end
+
+      ##
+      # Insert an hook after another specified hook
+      #
+      # @param [Class] after_class The hook to insert after
+      # @param [Class] hook_class The class of the hook to add
+      # @param [Hash] options A hash of options to pass into the hook during initialization
+      # @raise [HookNotFoundError] if the after hook is not found
+      #
+      def insert_after(after_class, hook_class, options = {})
+        hooks_mutex do
+          pos = @registry.find_index { |opts| opts.fetch(:klass, '') == after_class }
+          raise HookNotFoundError if pos.nil?
+
+          @registry.insert(
+            (pos + 1),
+            klass: hook_class,
+            options: options
+          )
+        end
+      end
+
+      ##
+      # Return a list of the hook classes in the registry in their execution order
+      #
+      # @return [Array<Class>]
+      #
+      def list
+        hooks_mutex do
+          @registry.map { |h| h[:klass] }
+        end
+      end
+
+      ##
+      # Lazily load and return all hooks for the given request
+      #
+      # @return [Array<Gruf::Hooks::Base>]
+      #
+      def prepare
+        is = []
+        hooks_mutex do
+          @registry.each do |o|
+            is << o[:klass].new(options: o[:options])
+          end
+        end
+        is
+      end
+
+      ##
+      # Clear the registry
+      #
+      def clear
+        hooks_mutex do
+          @registry = []
+        end
+      end
+
+      ##
+      # @return [Integer] The number of hooks currently loaded
+      #
+      def count
+        hooks_mutex do
+          @registry ||= []
+          @registry.count
+        end
+      end
+
+      private
+
+      ##
+      # Handle mutations to the hook registry in a thread-safe manner
+      #
+      def hooks_mutex(&block)
+        @hooks_mutex ||= begin
```

Comment:
I always was told `||=` is not thread safe, is that no longer the case? https://stackoverflow.com/questions/9853086/is-in-ruby-thread-safe

Reply:
Given that this just initializes the mutex if it's not initialized, which will happen before any threading takes place in the server, we should be fine here. It's also worth it to note this is assignment to a private variable in a private method, so unless someone is doing "something very bad" with `send`, it's not likely that will affect us here. (The interceptor registry uses the same exact approach, fwiw.)